<a href="https://colab.research.google.com/github/mrdbourke/pytorch-deep-learning/blob/main/00_pytorch_fundamentals.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a> 

[View Source Code](https://github.com/mrdbourke/pytorch-deep-learning/blob/main/00_pytorch_fundamentals.ipynb) | [View Slides](https://github.com/mrdbourke/pytorch-deep-learning/blob/main/slides/00_pytorch_and_deep_learning_fundamentals.pdf) | [Watch Video Walkthrough](https://youtu.be/Z_ikDlimN6A?t=76) 

# 1.3 PyTorch Grundlagen

- [PyTorch] (https://pytorch.org/) ist ein Open-Source-Framework für maschinelles Lernen und Deep Learning.
- Quelle: [Daniel Bourke](https://github.com/mrdbourke)

## Was ist PyTorch?
PyTorch ist ein Open-Source-Framework für maschinelles Lernen und Deep Learning. Mit PyTorch können Sie Daten manipulieren und verarbeiten und Algorithmen für maschinelles Lernen mit Python-Code schreiben. Viele der weltweit größten Technologieunternehmen wie [Meta (Facebook)](https://ai.facebook.com/blog/pytorch-builds-the-future-of-ai-and-machine-learning-at-facebook/), Tesla und Microsoft sowie Forschungsunternehmen für künstliche Intelligenz wie [OpenAI verwenden PyTorch](https://openai.com/blog/openai-pytorch/), um ihre Forschung voranzutreiben und maschinelles Lernen in ihre Produkte zu integrieren.

![pytorch wird in Industrie und Forschung eingesetzt](https://raw.githubusercontent.com/mrdbourke/pytorch-deep-learning/main/images/00-pytorch-being-used-across-research-and-industry.png)

Andrej Karpathy (Leiter der Abteilung für künstliche Intelligenz bei Tesla) hat beispielsweise in mehreren Vorträgen ([PyTorch DevCon 2019](https://youtu.be/oBklltKXtDE), [Tesla AI Day 2021](https://youtu.be/j0z4FweCy4M?t=2904)) darüber gesprochen, wie Tesla PyTorch nutzt, um seine selbstfahrenden Computer-Vision-Modelle zu betreiben.

PyTorch wird auch in anderen Branchen eingesetzt, z. B. in der Landwirtschaft, um [Computer Vision auf Traktoren zu betreiben](https://medium.com/pytorch/ai-for-ag-production-machine-learning-for-agriculture-e8cfdb9849a1).


## Warum PyTorch verwenden?
Forscher im Bereich des maschinellen Lernens verwenden PyTorch sehr gerne. Und im Februar 2022 ist PyTorch das [meistgenutzte Deep-Learning-Framework auf Papers With Code] (https://paperswithcode.com/trends), einer Website zur Verfolgung von Forschungsarbeiten zum maschinellen Lernen und den dazugehörigen Code-Repositories. PyTorch kümmert sich auch um viele Dinge wie GPU-Beschleunigung (damit Ihr Code schneller läuft) im Hintergrund. Sie können sich also auf die Bearbeitung von Daten und das Schreiben von Algorithmen konzentrieren, und PyTorch sorgt dafür, dass diese schnell ausgeführt werden. Und wenn Unternehmen wie Tesla und Meta (Facebook) es nutzen, um Modelle zu erstellen, mit denen sie Hunderte von Anwendungen betreiben, Tausende von Autos antreiben und Inhalte für Milliarden von Menschen bereitstellen, ist es eindeutig auch an der Entwicklungsfront fähig.

## Was wir in diesem Modul behandeln werden

| **Thema** | **Inhalt** |
| ----- | ----- |
| **Einführung in Tensoren** | Tensoren sind der Grundbaustein für maschinelles Lernen und Deep Learning. |
| **Erstellen von Tensoren** | Tensoren können fast jede Art von Daten darstellen (Bilder, Wörter, Zahlentabellen). |
| **Informationen aus Tensoren gewinnen** | Wenn man Informationen in einen Tensor eingeben kann, will man sie auch wieder herausbekommen. |
| **Manipulation von Tensoren** | Algorithmen für maschinelles Lernen (wie neuronale Netze) beinhalten die Manipulation von Tensoren auf viele verschiedene Arten, wie z. B. Addition, Multiplikation, Kombination. |
| Eines der häufigsten Probleme beim maschinellen Lernen ist der Umgang mit Formfehlern (der Versuch, falsch geformte Tensoren mit anderen Tensoren zu mischen). |
| **Indizierung auf Tensoren** | Wenn Sie schon einmal auf einer Python-Liste oder einem NumPy-Array indiziert haben, ist es bei Tensoren sehr ähnlich, außer dass sie viel mehr Dimensionen haben können. |
| PyTorch spielt mit Tensoren ([`torch.Tensor`](https://pytorch.org/docs/stable/tensors.html)), NumPy mag Arrays ([`np.ndarray`](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html)). Manchmal werden Sie diese mischen und anpassen wollen. |
| **Reproduzierbarkeit** | Maschinelles Lernen ist sehr experimentell und da es eine Menge *Zufälligkeit* verwendet, um zu funktionieren, wollen Sie manchmal, dass diese *Zufälligkeit* nicht so zufällig ist. |
| **Ausführen von Tensoren auf GPU** | GPUs (Graphics Processing Units) machen Ihren Code schneller, PyTorch macht es einfach, Ihren Code auf GPUs auszuführen. |

Alle Materialien für diesen Kurs [live auf GitHub] (https://github.com/mrdbourke/pytorch-deep-learning).
Und wenn Sie auf Probleme stoßen, können Sie dort auch eine Frage auf der [Diskussionsseite](https://github.com/mrdbourke/pytorch-deep-learning/discussions) stellen. Es gibt auch die [PyTorch-Entwicklerforen](https://discuss.pytorch.org/), ein sehr hilfreicher Ort für alles, was mit PyTorch zu tun hat.



## Importieren von PyTorch

> **Hinweis:** Bevor Sie den Code in diesem Notizbuch ausführen, sollten Sie die [PyTorch-Einrichtungsschritte] (https://pytorch.org/get-started/locally/) durchlaufen haben.
> **Wenn Sie jedoch mit Google Colab** arbeiten, sollte alles funktionieren (Google Colab wird mit PyTorch und anderen Bibliotheken installiert).

Beginnen wir damit, PyTorch zu importieren und die verwendete Version zu überprüfen.



In [ ]:
import torch

torch.__version__


'2.0.0+cpu'

Dies bedeutet, wenn Sie durch diese Materialien gehen, werden Sie sehen, die meisten Kompatibilität mit PyTorch 2.0.0 +, aber wenn Ihre Versionsnummer ist weit höher als das, könnten Sie einige Unstimmigkeiten bemerken.

Und wenn Sie irgendwelche Probleme haben, posten Sie bitte im Kurs [GitHub Discussions page] (https://github.com/mrdbourke/pytorch-deep-learning/discussions).

## Einführung in Tensoren

Nachdem wir nun PyTorch importiert haben, ist es an der Zeit, etwas über Tensoren zu lernen. Tensoren sind der grundlegende Baustein des maschinellen Lernens. Ihre Aufgabe ist es, Daten auf numerische Weise darzustellen.

Zum Beispiel könnte man ein Bild als Tensor mit der Form `[3, 224, 224]` darstellen, was `[Farbkanäle, Höhe, Breite]` bedeuten würde, denn das Bild hat `3` Farbkanäle (rot, grün, blau), eine Höhe von `224` Pixeln und eine Breite von `224` Pixeln.

![Beispiel für die Umwandlung eines Eingabebildes in eine Tensordarstellung des Bildes, das Bild wird in 3 Farbkanäle sowie in Zahlen zur Darstellung der Höhe und Breite zerlegt](https://raw.githubusercontent.com/mrdbourke/pytorch-deep-learning/main/images/00-tensor-shape-example-of-image.png)

In der Tensor-Sprache (der Sprache, die zur Beschreibung von Tensoren verwendet wird) hätte der Tensor drei Dimensionen, eine für "Farbkanäle", "Höhe" und "Breite".

### Tensoren erstellen

PyTorch liebt Tensoren. So sehr, dass es eine ganze Dokumentationsseite gibt, die der Klasse [`torch.Tensor`](https://pytorch.org/docs/stable/tensors.html) gewidmet ist. Ihre erste Hausaufgabe ist es, 10 Minuten lang [die Dokumentation zu `torch.Tensor`](https://pytorch.org/docs/stable/tensors.html) durchzulesen. Aber dazu können Sie später kommen. 

- Das erste, was wir erstellen werden, ist ein **Skalar**. 
- Ein Skalar ist eine einzelne Zahl, und in der Tensor-Sprache ist es ein Tensor mit null Dimension.


In [ ]:
# Scalar
scalar = torch.tensor(7)
scalar


tensor(7)

- Sehen Sie, wie oben `Tensor(7)` ausgedruckt wird? 
- Das bedeutet, obwohl `Skalar` eine einzelne Zahl ist, ist sie vom Typ `Torch.Tensor`.
- Wir können die Dimensionen eines Tensors mit dem Attribut `ndim` überprüfen.

In [ ]:
scalar.ndim


0

Was wäre, wenn wir die Zahl aus dem Tensor abrufen wollten?

- Zum Beispiel, indem wir sie von `torch.Tensor` in eine Python-Ganzzahl umwandeln?
- Dazu können wir die Methode `item()` verwenden.

In [ ]:
# Get the Python number within a tensor (only works with one-element tensors)
scalar.item()


7

Okay, sehen wir uns jetzt einen **Vektor** an.

- Ein Vektor ist ein eindimensionaler Tensor, kann aber viele Zahlen enthalten.
- Du könntest z.B. einen Vektor `[3, 2]` haben, um `[Schlafzimmer, Bäder]` in deinem Haus zu beschreiben. Oder du könntest `[3, 2, 2]` haben, um `[Schlafzimmer, Badezimmer, Parkplätze]` in deinem Haus zu beschreiben.
- Die wichtige Eigenschaft hier ist, dass ein Vektor flexibel ist in dem, was er darstellen kann (dasselbe gilt für Tensoren).



In [ ]:
# Vector
vector = torch.tensor([7, 7])
vector


tensor([7, 7])

Wunderbar, "Vektor" enthält jetzt zwei 7er, meine Lieblingszahl.

Wie viele Dimensionen wird er wohl haben?

In [ ]:
# Check the number of dimensions of vector
vector.ndim


1

Hmm, das ist seltsam, `vector` enthält zwei Zahlen, hat aber nur eine einzige Dimension.


- Sie können die Anzahl der Dimensionen eines Tensors in PyTorch an der Anzahl der eckigen Klammern auf der Außenseite (`[`) erkennen, und Sie brauchen nur eine Seite zu zählen.
- Wie viele eckige Klammern hat `vector`?
- Ein weiteres wichtiges Konzept für Tensoren ist ihr Attribut `shape`. Die Form gibt an, wie die Elemente in ihnen angeordnet sind.

Schauen wir uns die Form von `vector` an.



In [ ]:
# Check shape of vector
vector.shape


torch.Size([2])

Das obige Ergebnis ist `torch.Size([2])`, was bedeutet, dass unser Vektor eine Form von `[2]` hat. Das liegt an den beiden Elementen, die wir innerhalb der eckigen Klammern platziert haben (`[7, 7]`).

Sehen wir uns nun eine **Matrix** an.

In [ ]:
# Matrix
MATRIX = torch.tensor([[7, 8], [9, 10]])
MATRIX


tensor([[ 7,  8],
        [ 9, 10]])

Wow! Mehr Zahlen! Matrizen sind genauso flexibel wie Vektoren, nur dass sie eine zusätzliche Dimension haben.



In [ ]:
# Check number of dimensions
MATRIX.ndim


2

MATRIX" hat zwei Dimensionen (haben Sie die Anzahl der eckigen Klammern auf der Außenseite einer Seite gezählt?).

Welche "Form" wird er deiner Meinung nach haben?

In [ ]:
MATRIX.shape


torch.Size([2, 2])

Wir erhalten die Ausgabe `torch.Size([2, 2])`, weil `MATRIX` zwei Elemente tief und zwei Elemente breit ist.

Wie wäre es, wenn wir einen **Tensor** erstellen?

In [ ]:
# Tensor
TENSOR = torch.tensor([[[1, 2, 3], [3, 6, 9], [2, 4, 5]]])
TENSOR


tensor([[[1, 2, 3],
         [3, 6, 9],
         [2, 4, 5]]])

Wahnsinn! Was für ein gut aussehender Tensor.

Ich möchte betonen, dass Tensoren fast alles darstellen können.

Der Tensor, den wir gerade erstellt haben, könnte die Verkaufszahlen für ein Steak- und ein Mandelbuttergeschäft darstellen (zwei meiner Lieblingslebensmittel).

![ein einfacher Tensor in Google Sheets, der den Wochentag, die Steakverkäufe und die Mandelbutterverkäufe anzeigt](https://raw.githubusercontent.com/mrdbourke/pytorch-deep-learning/main/images/00_simple_tensor.png)

Wie viele Dimensionen hat er Ihrer Meinung nach? (Tipp: Verwenden Sie den Trick mit den eckigen Klammern)



In [ ]:
# Check number of dimensions for TENSOR
TENSOR.ndim


3

Und was ist mit seiner Form?

In [ ]:
# Check shape of TENSOR
TENSOR.shape


torch.Size([1, 3, 3])

In Ordnung, es wird `torch.Size([1, 3, 3])` ausgegeben.

- Die Dimensionen gehen von außen nach innen.
- Das bedeutet, dass es 1 Dimension von 3 mal 3 gibt.

<img src="https://raw.githubusercontent.com/mrdbourke/pytorch-deep-learning/main/images/00-pytorch-different-tensor-dimensions.png" alt="Beispiel für verschiedene Tensordimensionen" width="800">

> **Anmerkung:** Sie haben vielleicht bemerkt, dass ich Kleinbuchstaben für `Skalar` und `Vektor` und Großbuchstaben für `MATRIX` und `TENSOR` verwendet habe. Das war Absicht. In der Praxis sieht man Skalare und Vektoren oft als Kleinbuchstaben wie "y" oder "a". Und Matrizen und Tensoren werden mit Großbuchstaben wie "X" oder "W" bezeichnet.
>
> Sie werden auch feststellen, dass die Bezeichnungen "Matrix" und "Tensor" synonym verwendet werden. Das ist üblich. Da man es in PyTorch oft mit `torch.Tensor` zu tun hat (daher der Tensorname), bestimmen die Form und die Dimensionen dessen, was sich darin befindet, was es tatsächlich ist.

Lassen Sie uns zusammenfassen.

| Name | Was ist es? | Anzahl der Dimensionen | Untere oder obere (normalerweise/Beispiel) |
| ----- | ----- | ----- | ----- |
**Skalar** | eine einzelne Zahl | 0 | niedriger (`a`) |
| **Vektor** | eine Zahl mit Richtung (z. B. Windgeschwindigkeit mit Richtung), kann aber auch viele andere Zahlen enthalten | 1 | Lower (`y`) |
| **Matrix** | eine 2-dimensionale Anordnung von Zahlen | 2 | Upper (`Q`) |
**Tensor** | eine n-dimensionale Anordnung von Zahlen | kann eine beliebige Zahl sein, ein Tensor der Dimension 0 ist ein Skalar, ein Tensor der Dimension 1 ist ein Vektor | Upper (`X`) |





<img src="https://raw.githubusercontent.com/mrdbourke/pytorch-deep-learning/main/images/00-scalar-vector-matrix-tensor.png" alt="Skalar-Vektor-Matrix-Tensor und wie sie aussehen" width="600">


### Zufällige Tensoren

Wir haben festgestellt, dass Tensoren eine Form von Daten darstellen. Und maschinelle Lernmodelle wie neuronale Netze manipulieren und suchen Muster in Tensoren.
Bei der Erstellung von maschinellen Lernmodellen mit PyTorch ist es jedoch selten, dass man Tensoren von Hand erstellt (wie wir es getan haben). Stattdessen beginnt ein maschinelles Lernmodell oft mit großen zufälligen Zahlentensoren und passt diese Zufallszahlen an, während es die Daten durcharbeitet, um sie besser darzustellen. 
Im Wesentlichen bedeutet dies:
- "Mit Zufallszahlen beginnen -> Daten betrachten -> Zufallszahlen aktualisieren -> Daten betrachten -> Zufallszahlen aktualisieren...".

Als Datenwissenschaftler können Sie festlegen, wie das Modell für maschinelles Lernen startet (Initialisierung), die Daten betrachtet (Darstellung) und seine Zufallszahlen aktualisiert (Optimierung). Wir werden uns später mit diesen Schritten befassen. Sehen wir uns zunächst an, wie man einen Tensor mit Zufallszahlen erstellt.

Wir können dies mit [`torch.rand()`](https://pytorch.org/docs/stable/generated/torch.rand.html) und der Übergabe des Parameters `size` tun.


In [ ]:
# Create a random tensor of size (3, 4)
random_tensor = torch.rand(size=(3, 4))
random_tensor, random_tensor.dtype


(tensor([[0.6541, 0.4807, 0.2162, 0.6168],
         [0.4428, 0.6608, 0.6194, 0.8620],
         [0.2795, 0.6055, 0.4958, 0.5483]]),
 torch.float32)

Die Flexibilität von `torch.rand()` besteht darin, dass wir die `Größe so einstellen können, wie wir wollen. Nehmen wir zum Beispiel an, Sie wollten einen zufälligen Tensor in der üblichen Bildform von `[224, 224, 3]` (`[Höhe, Breite, Farbkanäle`]).

In [ ]:
# Create a random tensor of size (224, 224, 3)
random_image_size_tensor = torch.rand(size=(224, 224, 3))
random_image_size_tensor.shape, random_image_size_tensor.ndim


(torch.Size([224, 224, 3]), 3)

### Nullen und Einsen

Manchmal möchte man Tensoren einfach mit Nullen oder Einsen füllen. Dies geschieht häufig bei der Maskierung (z.B. Maskierung einiger Werte in einem Tensor mit Nullen, damit ein Modell weiß, dass es sie nicht lernen soll). Lassen Sie uns einen Tensor voller Nullen mit [`torch.zeros()`](https://pytorch.org/docs/stable/generated/torch.zeros.html) erstellen. Wieder kommt der Parameter `size` ins Spiel.


In [ ]:
# Create a tensor of all zeros
zeros = torch.zeros(size=(3, 4))
zeros, zeros.dtype


(tensor([[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]]),
 torch.float32)

Wir können dasselbe tun, um einen Tensor mit allen Einsen zu erzeugen, aber stattdessen [`torch.ones()` ](https://pytorch.org/docs/stable/generated/torch.ones.html) verwenden.

In [ ]:
# Create a tensor of all ones
ones = torch.ones(size=(3, 4))
ones, ones.dtype


(tensor([[1., 1., 1., 1.],
         [1., 1., 1., 1.],
         [1., 1., 1., 1.]]),
 torch.float32)

### Erstellen eines Bereichs und Tensoren wie

Manchmal braucht man einen Zahlenbereich, z. B. 1 bis 10 oder 0 bis 100.

Dazu können Sie `torch.arange(start, end, step)` verwenden.

Wobei:
* `Start` = Beginn des Bereichs (z.B. 0)
* `Ende` = Ende des Bereichs (z.B. 10)
* "step" = wie viele Schritte zwischen den einzelnen Werten (z.B. 1)

> **Anmerkung:** In Python können Sie `range()` verwenden, um einen Bereich zu erstellen. In PyTorch ist `torch.range()` jedoch veraltet und kann in Zukunft einen Fehler anzeigen.



In [ ]:
# Use torch.arange(), torch.range() is deprecated
zero_to_ten_deprecated = torch.arange(
    0, 10
)  # Note: this may return an error in the future

# Create a range of values 0 to 10
zero_to_ten = torch.arange(start=0, end=10, step=1)
zero_to_ten


tensor([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

- Manchmal möchte man einen Tensor eines bestimmten Typs mit der gleichen Form wie einen anderen Tensor haben.
- Zum Beispiel einen Tensor mit lauter Nullen, der die gleiche Form hat wie ein vorheriger Tensor.
- Dazu können Sie [`torch.zeros_like(input)`](https://pytorch.org/docs/stable/generated/torch.zeros_like.html) oder [`torch.ones_like(input)`](https://pytorch.org/docs/1.9.1/generated/torch.ones_like.html) verwenden, die einen Tensor mit Nullen bzw. Einsen in der gleichen Form wie der `input` zurückgeben.



In [ ]:
# Can also create a tensor of zeros similar to another tensor
ten_zeros = torch.zeros_like(input=zero_to_ten)  # will have same shape
ten_zeros


tensor([0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

### Tensor-Datentypen

Es gibt viele verschiedene [Tensor-Datentypen in PyTorch](https://pytorch.org/docs/stable/tensors.html#data-types).

- Einige sind spezifisch für die CPU und einige sind besser für die GPU geeignet. Es kann einige Zeit dauern, um herauszufinden, welche das sind.
- Wenn Sie irgendwo `torch.cuda` sehen, wird der Tensor im Allgemeinen für die GPU verwendet (da Nvidia GPUs ein Computing Toolkit namens CUDA verwenden).
- Der gebräuchlichste Typ (und im Allgemeinen der Standard) ist `torch.float32` oder `torch.float`. Dies wird als "32-Bit-Gleitkomma" bezeichnet.
- Aber es gibt auch 16-Bit-Gleitkomma (`torch.float16` oder `torch.half`) und 64-Bit-Gleitkomma (`torch.float64` oder `torch.double`). Und um die Dinge noch mehr zu verwirren, gibt es auch 8-Bit, 16-Bit, 32-Bit und 64-Bit Ganzzahlen.

**Anmerkung:** Ein Integer ist eine flache, runde Zahl wie `7`, während ein Float eine dezimale `7.0` hat. Der Grund dafür ist die **Präzision beim Rechnen**. Die Genauigkeit ist die Detailgenauigkeit, die zur Beschreibung einer Zahl verwendet wird. Je höher der Präzisionswert (8, 16, 32), desto mehr Details und damit Daten werden zur Beschreibung einer Zahl verwendet. Dies ist beim Deep Learning und bei der numerischen Datenverarbeitung von Bedeutung, denn es werden so viele Operationen durchgeführt, dass der Rechenaufwand umso höher ist, je detaillierter die Daten sind.

Datentypen mit geringerer Genauigkeit sind also in der Regel schneller in der Berechnung, haben aber Einbußen bei Bewertungsmetriken wie der Genauigkeit (schneller in der Berechnung, aber weniger genau).

**Ressourcen:**
  * Siehe die [PyTorch-Dokumentation für eine Liste aller verfügbaren Tensor-Datentypen](https://pytorch.org/docs/stable/tensors.html#data-types).
  * Lesen Sie die [Wikipedia-Seite für einen Überblick darüber, was Präzision in der Informatik](https://en.wikipedia.org/wiki/Precision_(computer_science)) ist.

Schauen wir uns an, wie man einige Tensoren mit bestimmten Datentypen erstellt. Wir können dies mit dem Parameter `dtype` tun.



In [ ]:
# Default datatype for tensors is float32
float_32_tensor = torch.tensor(
    [3.0, 6.0, 9.0],
    dtype=None,  # defaults to None, which is torch.float32 or whatever datatype is passed
    device=None,  # defaults to None, which uses the default tensor type
    requires_grad=False,
)  # if True, operations performed on the tensor are recorded

float_32_tensor.shape, float_32_tensor.dtype, float_32_tensor.device


(torch.Size([3]), torch.float32, device(type='cpu'))

Abgesehen von Problemen mit der Form (Tensorformen stimmen nicht überein), sind zwei der häufigsten Probleme, auf die man in PyTorch stößt, Datentyp- und Geräteprobleme. Zum Beispiel ist einer der Tensoren `torch.float32` und der andere `torch.float16` (PyTorch möchte oft, dass die Tensoren das gleiche Format haben). Oder einer Ihrer Tensoren befindet sich auf der CPU und der andere auf der GPU (PyTorch möchte, dass die Berechnungen zwischen den Tensoren auf demselben Gerät stattfinden).

Für den Moment wollen wir einen Tensor mit `dtype=torch.float16` erstellen.



In [ ]:
float_16_tensor = torch.tensor(
    [3.0, 6.0, 9.0], dtype=torch.float16
)  # torch.half would also work

float_16_tensor.dtype


torch.float16

## Informationen von Tensoren erhalten

Sobald Sie Tensoren erstellt haben (oder jemand anderes oder ein PyTorch-Modul sie für Sie erstellt hat), möchten Sie vielleicht einige Informationen von ihnen erhalten.

Wir haben diese bereits gesehen, aber drei der häufigsten Attribute, die man über Tensoren herausfinden möchte, sind:
* "Form" - welche Form hat der Tensor? (einige Operationen erfordern spezifische Formregeln)
* `dtype` - in welchem Datentyp sind die Elemente innerhalb des Tensors gespeichert?
* Gerät" - auf welchem Gerät ist der Tensor gespeichert? (normalerweise GPU oder CPU)

Lassen Sie uns einen zufälligen Tensor erstellen und Details über ihn herausfinden.


In [ ]:
# Create a tensor
some_tensor = torch.rand(3, 4)

# Find out details about it
print(some_tensor)
print(f"Shape of tensor: {some_tensor.shape}")
print(f"Datatype of tensor: {some_tensor.dtype}")
print(f"Device tensor is stored on: {some_tensor.device}")  # will default to CPU


tensor([[0.0079, 0.8656, 0.5640, 0.0424],
        [0.4597, 0.9129, 0.6073, 0.4909],
        [0.0170, 0.3657, 0.8348, 0.9291]])
Shape of tensor: torch.Size([3, 4])
Datatype of tensor: torch.float32
Device tensor is stored on: cpu


> **Hinweis:** Wenn Sie in PyTorch auf Probleme stoßen, hat das sehr oft mit einem der drei oben genannten Attribute zu tun. Wenn also die Fehlermeldungen auftauchen, singen Sie sich selbst ein kleines Lied namens "was, was, wo":
  * "*Welche Form haben meine Tensoren? welchen Datentyp haben sie und wo sind sie gespeichert? welche Form, welcher Datentyp, wo wo wo*"

## Manipulation von Tensoren (Tensoroperationen)

Beim Deep Learning werden Daten (Bilder, Text, Video, Audio, Proteinstrukturen usw.) als Tensoren dargestellt. Ein Modell lernt, indem es diese Tensoren untersucht und eine Reihe von Operationen (es könnten mehr als 1.000.000 sein) auf Tensoren ausführt, um eine Darstellung der Muster in den Eingabedaten zu erstellen.

Diese Operationen sind oft ein wunderbarer Tanz zwischen:
* Addition
* Subtraktion
* Multiplikation (elementweise)
* Division
* Matrix-Multiplikation

Und das war's. Sicher gibt es hier und da noch ein paar mehr, aber das sind die Grundbausteine neuronaler Netze. Stapelt man diese Bausteine auf die richtige Weise, kann man die ausgeklügeltsten neuronalen Netze erstellen (wie bei Lego!).



### Grundlegende Operationen

Beginnen wir mit einigen der grundlegenden Operationen: Addition (`+`), Subtraktion (`-`), Multiplikation (`*`).

Sie funktionieren genau so, wie Sie es sich vorstellen.

In [ ]:
# Create a tensor of values and add a number to it
tensor = torch.tensor([1, 2, 3])
tensor = tensor + 10
tensor


tensor([11, 12, 13])

In [40]:
# Multiply it by 10
tensor * 10


tensor([110, 120, 130])

Beachten Sie, dass die obigen Tensorwerte nicht als `tensor([110, 120, 130])` enden, weil sich die Werte innerhalb des Tensors nicht ändern, es sei denn, sie werden neu zugewiesen.

In [ ]:
# Tensors don't change unless reassigned
tensor


tensor([11, 12, 13])

Wir subtrahieren eine Zahl und weisen diesmal die Variable `Tensor` neu zu.

In [ ]:
# Subtract and reassign
tensor = tensor - 10
tensor


tensor([1, 2, 3])

In [ ]:
# Add and reassign
tensor = tensor + 10
tensor


tensor([11, 12, 13])

PyTorch hat auch eine Reihe von eingebauten Funktionen wie [`torch.mul()`](https://pytorch.org/docs/stable/generated/torch.mul.html#torch.mul) (kurz für Multiplikation) und [`torch.add()`](https://pytorch.org/docs/stable/generated/torch.add.html), um grundlegende Operationen durchzuführen.

In [ ]:
# Can also use torch functions
torch.multiply(tensor, 10)


tensor([110, 120, 130])

In [ ]:
# Original tensor is still unchanged
tensor


tensor([11, 12, 13])

Es ist jedoch gebräuchlicher, die Operator-Symbole wie "*" anstelle von "torch.mul()" zu verwenden.

In [ ]:
# Element-wise multiplication (each element multiplies its equivalent, index 0->0, 1->1, 2->2)
print(tensor, "*", tensor)
print("Equals:", tensor * tensor)


tensor([11, 12, 13]) * tensor([11, 12, 13])
Equals: tensor([121, 144, 169])


### Matrixmultiplikation (ist alles, was Sie brauchen)

Eine der gebräuchlichsten Operationen in Algorithmen für maschinelles Lernen und Deep Learning (wie neuronale Netze) ist die [Matrixmultiplikation](https://www.mathsisfun.com/algebra/matrix-multiplying.html).

PyTorch implementiert die Funktionalität der Matrixmultiplikation in der Methode [`torch.matmul()`](https://pytorch.org/docs/stable/generated/torch.matmul.html).

Die zwei wichtigsten Regeln für die Matrixmultiplikation, die man sich merken sollte, sind:

1. Die **inneren Dimensionen** müssen übereinstimmen:
  * `(3, 2) @ (3, 2)` wird nicht funktionieren
  * `(2, 3) @ (3, 2)` wird funktionieren
  * `(3, 2) @ (2, 3)` wird funktionieren
2. Die resultierende Matrix hat die Form der **Außendimensionen**:
 * `(2, 3) @ (3, 2)` -> `(2, 2)`
 * `(3, 2) @ (2, 3)` -> `(3, 3)`

> **Hinweis:** "`@`" ist in Python das Symbol für die Matrixmultiplikation.

> **Ressource:** Sie können alle Regeln für die Matrixmultiplikation mit `torch.matmul()` [in der PyTorch-Dokumentation] (https://pytorch.org/docs/stable/generated/torch.matmul.html) sehen.

Erstellen wir einen Tensor und führen wir eine elementweise Multiplikation und eine Matrixmultiplikation mit ihm durch.




In [ ]:
import torch

tensor = torch.tensor([1, 2, 3])
tensor.shape


torch.Size([3])

Der Unterschied zwischen der elementweisen Multiplikation und der Matrixmultiplikation besteht in der Addition von Werten.

Für unsere `Tensor`-Variable mit den Werten `[1, 2, 3]`:

| Operation | Berechnung | Code |
| ----- | ----- | ----- |
| **Elementweise Multiplikation** | `[1*1, 2*2, 3*3]` = `[1, 4, 9]` | `Tensor * Tensor` |
| **Matrix-Multiplikation** | `[1*1 + 2*2 + 3*3]` = `[14]` | `tensor.matmul(tensor)` |



In [ ]:
# Element-wise matrix multiplication
tensor * tensor


tensor([1, 4, 9])

In [ ]:
# Matrix multiplication
torch.matmul(tensor, tensor)


tensor(14)

In [ ]:
# Can also use the "@" symbol for matrix multiplication, though not recommended
tensor @ tensor


tensor(14)

Sie können die Matrixmultiplikation von Hand durchführen, aber das ist nicht empfehlenswert. Die eingebaute Methode `torch.matmul()` ist schneller.

In [ ]:
%%time
# Matrix multiplication by hand
# (avoid doing operations with for loops at all cost, they are computationally expensive)
value = 0
for i in range(len(tensor)):
    value += tensor[i] * tensor[i]
value


CPU times: total: 0 ns
Wall time: 1 ms


tensor(14)

In [ ]:
%%time
torch.matmul(tensor, tensor)


CPU times: total: 0 ns
Wall time: 0 ns


tensor(14)

## Einer der häufigsten Fehler beim Deep Learning (Formfehler)

Da ein großer Teil des Deep Learning aus der Multiplikation und der Durchführung von Operationen mit Matrizen besteht und für Matrizen strenge Regeln gelten, welche Formen und Größen kombiniert werden können, ist einer der häufigsten Fehler, auf den man beim Deep Learning stößt, die Nichtübereinstimmung von Formen.

In [ ]:
# Shapes need to be in the right way
tensor_A = torch.tensor([[1, 2], [3, 4], [5, 6]], dtype=torch.float32)

tensor_B = torch.tensor([[7, 10], [8, 11], [9, 12]], dtype=torch.float32)

torch.matmul(tensor_A, tensor_B)  # (this will error)


RuntimeError: mat1 and mat2 shapes cannot be multiplied (3x2 and 3x2)

Wir können die Matrixmultiplikation zwischen "Tensor_A" und "Tensor_B" durchführen, indem wir ihre inneren Dimensionen übereinstimmen lassen.

Eine der Möglichkeiten, dies zu tun, ist eine **Transposition** (die Dimensionen eines gegebenen Tensors vertauschen).

Sie können Transpositionen in PyTorch durchführen, indem Sie entweder:
* `torch.transpose(input, dim0, dim1)` - wobei `input` der gewünschte zu transponierende Tensor ist und `dim0` und `dim1` die zu vertauschenden Dimensionen sind.
* `tensor.T` - wobei `tensor` der gewünschte Tensor ist, der transponiert werden soll.

Versuchen wir Letzteres.

In [ ]:
# View tensor_A and tensor_B
print(tensor_A)
print(tensor_B)


tensor([[1., 2.],
        [3., 4.],
        [5., 6.]])
tensor([[ 7., 10.],
        [ 8., 11.],
        [ 9., 12.]])


In [ ]:
# View tensor_A and tensor_B.T
print(tensor_A)
print(tensor_B.T)


tensor([[1., 2.],
        [3., 4.],
        [5., 6.]])
tensor([[ 7.,  8.,  9.],
        [10., 11., 12.]])


In [ ]:
# The operation works when tensor_B is transposed
print(f"Original shapes: tensor_A = {tensor_A.shape}, tensor_B = {tensor_B.shape}\n")
print(
    f"New shapes: tensor_A = {tensor_A.shape} (same as above), tensor_B.T = {tensor_B.T.shape}\n"
)
print(f"Multiplying: {tensor_A.shape} * {tensor_B.T.shape} <- inner dimensions match\n")
print("Output:\n")
output = torch.matmul(tensor_A, tensor_B.T)
print(output)
print(f"\nOutput shape: {output.shape}")


Original shapes: tensor_A = torch.Size([3, 2]), tensor_B = torch.Size([3, 2])

New shapes: tensor_A = torch.Size([3, 2]) (same as above), tensor_B.T = torch.Size([2, 3])

Multiplying: torch.Size([3, 2]) * torch.Size([2, 3]) <- inner dimensions match

Output:

tensor([[ 27.,  30.,  33.],
        [ 61.,  68.,  75.],
        [ 95., 106., 117.]])

Output shape: torch.Size([3, 3])


Sie können auch [`torch.mm()`](https://pytorch.org/docs/stable/generated/torch.mm.html) verwenden, was eine Abkürzung für `torch.matmul()` ist.

In [ ]:
# torch.mm is a shortcut for matmul
torch.mm(tensor_A, tensor_B.T)


tensor([[ 27.,  30.,  33.],
        [ 61.,  68.,  75.],
        [ 95., 106., 117.]])

Ohne die Transponierung sind die Regeln der Matrixmultiplikation nicht erfüllt, und wir erhalten einen Fehler wie oben.

Wie wäre es mit einer visuellen Darstellung?

![Visuelle Demo der Matrixmultiplikation](https://github.com/mrdbourke/pytorch-deep-learning/raw/main/images/00-matrix-multiply-crop.gif)

Sie können Ihre eigenen visuellen Darstellungen der Matrixmultiplikation wie diese unter http://matrixmultiplication.xyz/ erstellen.

**Anmerkung:** Eine Matrixmultiplikation wie diese wird auch als das [**Punktprodukt**](https://www.mathsisfun.com/algebra/vectors-dot-product.html) zweier Matrizen bezeichnet.


Neuronale Netze sind voll von Matrixmultiplikationen und Skalarprodukten. Das Modul [`torch.nn.Linear()`](https://pytorch.org/docs/1.9.1/generated/torch.nn.Linear.html) (wir werden es später in Aktion sehen), auch bekannt als Feed-Forward-Schicht oder vollständig verbundene Schicht, implementiert eine Matrixmultiplikation zwischen einer Eingabe `x` und einer Gewichtungsmatrix `A`.

$$
y = W^{\top} \cdot x + b
$$

Wobei:
- $x$ ist die Eingabe für die Schicht (Deep Learning ist ein Stapel von Schichten wie `torch.nn.Linear()` und andere übereinander).
- $W$ ist die Gewichtungsmatrix, die von der Schicht erstellt wird. Am Anfang sind es Zufallszahlen, die angepasst werden, wenn ein neuronales Netzwerk lernt, Muster in den Daten besser darzustellen (beachten Sie das "T", das ist, weil die Gewichtungsmatrix transponiert wird). Oft wird auch $A$ oder ein anderer Buchstabe wie $\theta$ verwendet, um die Gewichtungsmatrix darzustellen.
- $b$ ist der Bias-Begriff, der verwendet wird, um die Gewichte und Eingaben leicht zu verschieben.
- $y$ ist die Ausgabe (eine Manipulation der Eingabe in der Hoffnung, Muster darin zu entdecken).

Dies ist eine lineare Funktion (vielleicht haben Sie in der Schule oder anderswo schon einmal etwas wie $y = mx+b$ gesehen) und kann zum Zeichnen einer geraden Linie verwendet werden!
- Versuchen Sie, die Werte von `in_features` und `out_features` zu ändern und sehen Sie, was passiert.
- Fällt Ihnen etwas auf, was mit den Formen zu tun hat?



In [ ]:
# Since the linear layer starts with a random weights matrix, let's make it reproducible (more on this later)
torch.manual_seed(42)
# This uses matrix multiplication
linear = torch.nn.Linear(
    in_features=2,  # in_features = matches inner dimension of input
    out_features=6,
)  # out_features = describes outer value
x = tensor_A
output = linear(x)
print(f"Input shape: {x.shape}\n")
print(f"Output:\n{output}\n\nOutput shape: {output.shape}")


Input shape: torch.Size([3, 2])

Output:
tensor([[2.2368, 1.2292, 0.4714, 0.3864, 0.1309, 0.9838],
        [4.4919, 2.1970, 0.4469, 0.5285, 0.3401, 2.4777],
        [6.7469, 3.1648, 0.4224, 0.6705, 0.5493, 3.9716]],
       grad_fn=<AddmmBackward0>)

Output shape: torch.Size([3, 6])


> **Question:** What happens if you change `in_features` from 2 to 3 above? Does it error? How could you change the shape of the input (`x`) to accommodate to the error? Hint: what did we have to do to `tensor_B` above?

Wenn Sie sich noch nie mit der Matrixmultiplikation beschäftigt haben, kann das Thema zunächst verwirrend sein.

Aber nachdem Sie ein paar Mal damit herumgespielt und sogar ein paar neuronale Netze geknackt haben, werden Sie feststellen, dass sie überall vorkommt.

Denken Sie daran: Matrixmultiplikation ist alles, was Sie brauchen.

![Matrixmultiplikation ist alles, was du brauchst](https://raw.githubusercontent.com/mrdbourke/pytorch-deep-learning/main/images/00_matrix_multiplication_is_all_you_need.jpeg)

*Wenn du anfängst, dich mit neuronalen Netzwerkschichten zu beschäftigen und deine eigenen zu bauen, wirst du überall Matrixmultiplikationen finden. **Quelle:** https://marksaroufim.substack.com/p/working-class-deep-learner*



### Min., Max., Mittelwert, Summe usw. ermitteln (Aggregation)

Nachdem wir nun einige Möglichkeiten zur Manipulation von Tensoren kennengelernt haben, wollen wir nun einige Möglichkeiten zur Aggregation von Tensoren durchspielen (von mehr Werten zu weniger Werten). Zuerst werden wir einen Tensor erstellen und dann das Maximum, das Minimum, den Mittelwert und die Summe davon finden.





In [ ]:
# Create a tensor
x = torch.arange(0, 100, 10)
x


tensor([ 0, 10, 20, 30, 40, 50, 60, 70, 80, 90])

Lassen Sie uns nun eine Aggregation durchführen.

In [ ]:
print(f"Minimum: {x.min()}")
print(f"Maximum: {x.max()}")
# print(f"Mean: {x.mean()}") # this will error
print(f"Mean: {x.type(torch.float32).mean()}")  # won't work without float datatype
print(f"Sum: {x.sum()}")


Minimum: 0
Maximum: 90
Mean: 45.0
Sum: 450


**Hinweis:** Einige Methoden, wie z.B. `torch.mean()`, erfordern, dass die Tensoren im Datentyp `torch.float32` (der häufigste) oder einem anderen spezifischen Datentyp vorliegen, andernfalls schlägt die Operation fehl.

Sie können auch das Gleiche wie oben mit `torch`-Methoden tun.

In [ ]:
torch.max(x), torch.min(x), torch.mean(x.type(torch.float32)), torch.sum(x)


(tensor(90), tensor(0), tensor(45.), tensor(450))

### Positionsbezogenes Min/Max

Sie können auch den Index eines Tensors, in dem das Maximum oder Minimum auftritt, mit [`torch.argmax()`](https://pytorch.org/docs/stable/generated/torch.argmax.html) bzw. [`torch.argmin()`](https://pytorch.org/docs/stable/generated/torch.argmin.html) ermitteln.

Dies ist hilfreich, wenn Sie nur die Position suchen, an der der höchste (oder niedrigste) Wert liegt, und nicht den eigentlichen Wert selbst (wir werden dies in einem späteren Abschnitt sehen, wenn wir die [softmax-Aktivierungsfunktion](https://pytorch.org/docs/stable/generated/torch.nn.Softmax.html) verwenden).



In [ ]:
# Create a tensor
tensor = torch.arange(10, 100, 10)
print(f"Tensor: {tensor}")

# Returns index of max and min values
print(f"Index where max value occurs: {tensor.argmax()}")
print(f"Index where min value occurs: {tensor.argmin()}")


Tensor: tensor([10, 20, 30, 40, 50, 60, 70, 80, 90])
Index where max value occurs: 8
Index where min value occurs: 0


### Tensor-Datentyp ändern

Wie bereits erwähnt, besteht ein häufiges Problem bei Deep-Learning-Operationen darin, dass die Tensoren in verschiedenen Datentypen vorliegen. Wenn ein Tensor im Datentyp `torch.float64` und ein anderer im Datentyp `torch.float32` ist, kann es zu Fehlern kommen. Aber es gibt eine Lösung.

- Sie können die Datentypen von Tensoren mit [`torch.Tensor.type(dtype=None)`](https://pytorch.org/docs/stable/generated/torch.Tensor.type.html) ändern, wobei der Parameter `dtype` der Datentyp ist, den Sie verwenden möchten.
- Zuerst erstellen wir einen Tensor und prüfen seinen Datentyp (die Vorgabe ist `torch.float32`).


In [ ]:
# Create a tensor and check its datatype
tensor = torch.arange(10.0, 100.0, 10.0)
tensor.dtype


torch.float32

Jetzt erstellen wir einen weiteren Tensor wie zuvor, ändern aber den Datentyp in "torch.float16".


In [ ]:
# Create a float16 tensor
tensor_float16 = tensor.type(torch.float16)
tensor_float16


tensor([10., 20., 30., 40., 50., 60., 70., 80., 90.], dtype=torch.float16)

Und wir können etwas Ähnliches tun, um einen `torch.int8`-Tensor zu erstellen.

In [ ]:
# Create an int8 tensor
tensor_int8 = tensor.type(torch.int8)
tensor_int8


tensor([10, 20, 30, 40, 50, 60, 70, 80, 90], dtype=torch.int8)

> **Anmerkung:** Verschiedene Datentypen können anfangs verwirrend sein. Je niedriger die Zahl ist (z. B. 32, 16, 8), desto ungenauer speichert ein Computer den Wert. Und eine geringere Speichermenge führt im Allgemeinen zu einer schnelleren Berechnung und einem kleineren Gesamtmodell. Mobile neuronale Netze arbeiten oft mit 8-Bit-Ganzzahlen, die kleiner und schneller sind, aber weniger genau als ihre Float32-Gegenstücke. Mehr dazu finden Sie in [precision in computing] (https://en.wikipedia.org/wiki/Precision_(computer_science)).

> **Übung:** Bis jetzt haben wir einige Tensor-Methoden behandelt, aber es gibt noch eine Menge mehr in der [`torch.Tensor` Dokumentation](https://pytorch.org/docs/stable/tensors.html), ich würde empfehlen, 10 Minuten damit zu verbringen, durchzublättern und sich alle anzuschauen, die Ihnen ins Auge fallen. Klicken Sie sie an und schreiben Sie sie dann selbst in Code, um zu sehen, was passiert.



### Umformen, Stapeln, Quetschen und Aufheben des Quetschens

Oft möchte man die Form oder die Dimensionen seiner Tensoren ändern, ohne die Werte darin zu verändern.

Hierfür gibt es einige beliebte Methoden:

| Methode | Einzeilige Beschreibung |
| ----- | ----- |
| [`torch.reshape(input, shape)`](https://pytorch.org/docs/stable/generated/torch.reshape.html#torch.reshape) | Formt `input` in `shape` um (falls kompatibel), kann auch `torch.Tensor.reshape()` verwenden. |
| [`Tensor.view(shape)`](https://pytorch.org/docs/stable/generated/torch.Tensor.view.html) | Gibt eine Ansicht des ursprünglichen Tensors in einer anderen `Form` zurück, hat aber die gleichen Daten wie der ursprüngliche Tensor. |
| [`torch.stack(tensors, dim=0)`](https://pytorch.org/docs/1.9.1/generated/torch.stack.html) | Verkettet eine Folge von `Tensoren` entlang einer neuen Dimension (`dim`), alle `Tensoren` müssen gleich groß sein. |
| [`torch.squeeze(input)`](https://pytorch.org/docs/stable/generated/torch.squeeze.html) | Quetscht `input` um alle Dimenionen mit dem Wert `1` zu entfernen. |
| [`torch.unsqueeze(input, dim)`](https://pytorch.org/docs/1.9.1/generated/torch.unsqueeze.html) | Gibt `input` mit einem Dimensionswert von `1` zurück, der bei `dim` hinzugefügt wurde. |
| [`torch.permute(input, dims)`](https://pytorch.org/docs/stable/generated/torch.permute.html) | Gibt eine *Ansicht* der ursprünglichen `Eingabe` mit ihren Dimensionen permutiert (umgeordnet) zu `dims` zurück. |

Warum dies alles tun?

Weil es bei Deep-Learning-Modellen (neuronalen Netzen) immer darum geht, Tensoren auf irgendeine Weise zu manipulieren. Und aufgrund der Regeln der Matrixmultiplikation kommt es zu Fehlern, wenn die Formen nicht übereinstimmen. Mit diesen Methoden können Sie sicherstellen, dass die richtigen Elemente Ihrer Tensoren mit den richtigen Elementen anderer Tensoren gemischt werden.

Probieren wir sie aus. Zuerst erstellen wir einen Tensor.



In [ ]:
# Create a tensor
import torch

x = torch.arange(1.0, 8.0)
x, x.shape


(tensor([1., 2., 3., 4., 5., 6., 7.]), torch.Size([7]))

Nun wollen wir mit `torch.reshape()` eine zusätzliche Dimension hinzufügen.

In [ ]:
# Add an extra dimension
x_reshaped = x.reshape(1, 7)
x_reshaped, x_reshaped.shape


(tensor([[1., 2., 3., 4., 5., 6., 7.]]), torch.Size([1, 7]))

Wir können auch die Ansicht mit `torch.view()` ändern.

In [ ]:
# Change view (keeps same data as original but changes view)
# See more: https://stackoverflow.com/a/54507446/7900723
z = x.view(1, 7)
z, z.shape


(tensor([[1., 2., 3., 4., 5., 6., 7.]]), torch.Size([1, 7]))

Denken Sie aber daran, dass das Ändern der Ansicht eines Tensors mit `torch.view()` wirklich nur eine neue Ansicht des *gleichen* Tensors erzeugt.
Das Ändern der Ansicht ändert also auch den ursprünglichen Tensor.

In [ ]:
# Changing z changes x
z[:, 0] = 5
z, x


(tensor([[5., 2., 3., 4., 5., 6., 7.]]), tensor([5., 2., 3., 4., 5., 6., 7.]))

Wenn wir unseren neuen Tensor fünfmal auf sich selbst stapeln wollten, könnten wir dies mit `torch.stack()` tun.

In [ ]:
# Stack tensors on top of each other
x_stacked = torch.stack(
    [x, x, x, x], dim=0
)  # try changing dim to dim=1 and see what happens
x_stacked


tensor([[5., 2., 3., 4., 5., 6., 7.],
        [5., 2., 3., 4., 5., 6., 7.],
        [5., 2., 3., 4., 5., 6., 7.],
        [5., 2., 3., 4., 5., 6., 7.]])

Wie wäre es, alle einzelnen Dimensionen aus einem Tensor zu entfernen?

Dazu können Sie `torch.squeeze()` verwenden (ich erinnere mich daran, dass der Tensor nur Dimensionen über 1 haben soll).

In [ ]:
print(f"Previous tensor: {x_reshaped}")
print(f"Previous shape: {x_reshaped.shape}")

# Remove extra dimension from x_reshaped
x_squeezed = x_reshaped.squeeze()
print(f"\nNew tensor: {x_squeezed}")
print(f"New shape: {x_squeezed.shape}")


Previous tensor: tensor([[5., 2., 3., 4., 5., 6., 7.]])
Previous shape: torch.Size([1, 7])

New tensor: tensor([5., 2., 3., 4., 5., 6., 7.])
New shape: torch.Size([7])


Und um das Gegenteil von "torch.squeeze()" zu tun, können Sie "torch.unsqueeze()" verwenden, um einen Dimensionswert von 1 bei einem bestimmten Index hinzuzufügen.

In [ ]:
print(f"Previous tensor: {x_squeezed}")
print(f"Previous shape: {x_squeezed.shape}")

## Add an extra dimension with unsqueeze
x_unsqueezed = x_squeezed.unsqueeze(dim=0)
print(f"\nNew tensor: {x_unsqueezed}")
print(f"New shape: {x_unsqueezed.shape}")


Previous tensor: tensor([5., 2., 3., 4., 5., 6., 7.])
Previous shape: torch.Size([7])

New tensor: tensor([[5., 2., 3., 4., 5., 6., 7.]])
New shape: torch.Size([1, 7])


Sie können auch die Reihenfolge der Achsenwerte mit `torch.permute(input, dims)` neu anordnen, wobei der `input` in eine *Ansicht* mit neuen `dims` umgewandelt wird.

In [ ]:
# Create tensor with specific shape
x_original = torch.rand(size=(224, 224, 3))

# Permute the original tensor to rearrange the axis order
x_permuted = x_original.permute(2, 0, 1)  # shifts axis 0->1, 1->2, 2->0

print(f"Previous shape: {x_original.shape}")
print(f"New shape: {x_permuted.shape}")


Previous shape: torch.Size([224, 224, 3])
New shape: torch.Size([3, 224, 224])


> **Anmerkung**: Da die Permutation eine *view* (mit denselben Daten wie das Original) ergibt, sind die Werte im permutierten Tensor dieselben wie im ursprünglichen Tensor, und wenn Sie die Werte in der Ansicht ändern, werden auch die Werte des Originals geändert.

## Indizierung (Auswahl von Daten aus Tensoren)

- Manchmal möchte man bestimmte Daten aus Tensoren auswählen (zum Beispiel nur die erste Spalte oder die zweite Zeile).
- Zu diesem Zweck können Sie die Indizierung verwenden.
- Wenn Sie schon einmal mit Python-Listen oder NumPy-Arrays indiziert haben, ist die Indizierung in PyTorch mit Tensoren sehr ähnlich.

In [ ]:
# Create a tensor
import torch

x = torch.arange(1, 10).reshape(1, 3, 3)
x, x.shape


(tensor([[[1, 2, 3],
          [4, 5, 6],
          [7, 8, 9]]]),
 torch.Size([1, 3, 3]))

Die Indizierung der Werte geht von der äußeren Dimension zur inneren Dimension (siehe die eckigen Klammern).

In [ ]:
# Let's index bracket by bracket
print(f"First square bracket:\n{x[0]}")
print(f"Second square bracket: {x[0][0]}")
print(f"Third square bracket: {x[0][0][0]}")


First square bracket:
tensor([[1, 2, 3],
        [4, 5, 6],
        [7, 8, 9]])
Second square bracket: tensor([1, 2, 3])
Third square bracket: 1


Sie können auch `:` verwenden, um "alle Werte in dieser Dimension" anzugeben und dann ein Komma (`,`) verwenden, um eine weitere Dimension hinzuzufügen.

In [ ]:
# Get all values of 0th dimension and the 0 index of 1st dimension
x[:, 0]


tensor([[1, 2, 3]])

In [ ]:
# Get all values of 0th & 1st dimensions but only index 1 of 2nd dimension
x[:, :, 1]


tensor([[2, 5, 8]])

In [ ]:
# Get all values of the 0 dimension but only the 1 index value of the 1st and 2nd dimension
x[:, 1, 1]


tensor([5])

In [ ]:
# Get index 0 of 0th and 1st dimension and all values of 2nd dimension
x[0, 0, :]  # same as x[0][0]


tensor([1, 2, 3])

Die Indizierung kann anfangs ziemlich verwirrend sein, vor allem bei größeren Tensoren (ich muss die Indizierung immer noch mehrmals ausprobieren, um sie richtig hinzubekommen). Aber mit ein bisschen Übung und dem Motto des Datenexplorers (***Visualisieren, visualisieren, visualisieren***) werden Sie den Dreh schon raus haben.

## PyTorch Tensoren & NumPy

Da NumPy eine beliebte Python-Bibliothek für numerische Berechnungen ist, verfügt PyTorch über Funktionen, die eine gute Interaktion mit ihr ermöglichen.

Die beiden wichtigsten Methoden, die Sie für den Übergang von NumPy zu PyTorch (und zurück) verwenden sollten, sind:
* [`torch.from_numpy(ndarray)`](https://pytorch.org/docs/stable/generated/torch.from_numpy.html) - NumPy-Array -> PyTorch-Tensor.
* [`torch.Tensor.numpy()`](https://pytorch.org/docs/stable/generated/torch.Tensor.numpy.html) - PyTorch Tensor -> NumPy Array.

Probieren wir sie aus.

In [ ]:
# NumPy array to tensor
import torch
import numpy as np

array = np.arange(1.0, 8.0)
tensor = torch.from_numpy(array)
array, tensor


(array([1., 2., 3., 4., 5., 6., 7.]),
 tensor([1., 2., 3., 4., 5., 6., 7.], dtype=torch.float64))

> **Hinweis:** Standardmäßig werden NumPy-Arrays mit dem Datentyp "Float64" erstellt, und wenn Sie sie in einen PyTorch-Tensor konvertieren, behalten sie denselben Datentyp (wie oben).
> Viele PyTorch-Berechnungen verwenden jedoch standardmäßig den Datentyp `Float32`.
> Wenn Sie also Ihr NumPy-Array (`float64`) -> PyTorch-Tensor (`float64`) -> PyTorch-Tensor (float32) konvertieren wollen, können Sie `tensor = torch.from_numpy(array).type(torch.float32)` verwenden.

Da wir `tensor` oben neu zugewiesen haben, bleibt das Array gleich, wenn Sie den Tensor ändern.



In [ ]:
# Change the array, keep the tensor
array = array + 1
array, tensor


(array([2., 3., 4., 5., 6., 7., 8.]),
 tensor([1., 2., 3., 4., 5., 6., 7.], dtype=torch.float64))

Und wenn Sie von einem PyTorch-Tensor zu einem NumPy-Array wechseln wollen, können Sie "tensor.numpy()" aufrufen.

In [ ]:
# Tensor to NumPy array
tensor = torch.ones(7)  # create a tensor of ones with dtype=float32
numpy_tensor = tensor.numpy()  # will be dtype=float32 unless changed
tensor, numpy_tensor


(tensor([1., 1., 1., 1., 1., 1., 1.]),
 array([1., 1., 1., 1., 1., 1., 1.], dtype=float32))

Und es gilt die gleiche Regel wie oben: Wenn Sie den ursprünglichen `Tensor` ändern, bleibt der neue `numpy_tensor` derselbe.

In [ ]:
# Change the tensor, keep the array the same
tensor = tensor + 1
tensor, numpy_tensor


(tensor([2., 2., 2., 2., 2., 2., 2.]),
 array([1., 1., 1., 1., 1., 1., 1.], dtype=float32))

## Reproduzierbarkeit (Versuch, den Zufall aus dem Zufall herauszuhalten)

Wenn Sie mehr über neuronale Netze und maschinelles Lernen lernen, werden Sie feststellen, wie sehr der Zufall eine Rolle spielt. Nun ja, eher der Pseudozufall, denn schließlich ist ein Computer grundsätzlich deterministisch (jeder Schritt ist vorhersehbar), so dass die Zufälligkeiten, die er erzeugt, simulierte Zufälligkeiten sind.

Was hat das nun mit neuronalen Netzen und Deep Learning zu tun? Wir haben besprochen, dass neuronale Netze mit Zufallszahlen beginnen, um Muster in Daten zu beschreiben (diese Zahlen sind schlechte Beschreibungen), und versuchen, diese Zufallszahlen mithilfe von Tensoroperationen (und einigen anderen Dingen, die wir noch nicht besprochen haben) zu verbessern, um Muster in Daten besser zu beschreiben.

Kurz gesagt:

``Anfang mit Zufallszahlen -> Tensoroperationen -> Versuch, es besser zu machen (wieder und wieder und wieder)``

Obwohl Zufälligkeit schön und mächtig ist, hätte man manchmal gerne etwas weniger Zufälligkeit. Und warum? Damit Sie wiederholbare Experimente durchführen können.

- Ein Beispiel: Sie entwickeln einen Algorithmus, der eine Leistung von X erreicht.
- Und dann probiert Ihr Freund ihn aus, um zu überprüfen, ob Sie nicht verrückt sind. Wie könnten sie so etwas tun?

Hier kommt die **Reproduzierbarkeit** ins Spiel. Mit anderen Worten: Können Sie auf Ihrem Computer mit demselben Code die gleichen (oder sehr ähnliche) Ergebnisse erzielen wie ich auf meinem? Schauen wir uns ein kurzes Beispiel für Reproduzierbarkeit in PyTorch an.

Wir beginnen damit, zwei zufällige Tensoren zu erzeugen. Da sie zufällig sind, sollte man erwarten, dass sie unterschiedlich sind, oder?



In [ ]:
import torch

# Create two random tensors
random_tensor_A = torch.rand(3, 4)
random_tensor_B = torch.rand(3, 4)

print(f"Tensor A:\n{random_tensor_A}\n")
print(f"Tensor B:\n{random_tensor_B}\n")
print(f"Does Tensor A equal Tensor B? (anywhere)")
random_tensor_A == random_tensor_B


Tensor A:
tensor([[0.8016, 0.3649, 0.6286, 0.9663],
        [0.7687, 0.4566, 0.5745, 0.9200],
        [0.3230, 0.8613, 0.0919, 0.3102]])

Tensor B:
tensor([[0.9536, 0.6002, 0.0351, 0.6826],
        [0.3743, 0.5220, 0.1336, 0.9666],
        [0.9754, 0.8474, 0.8988, 0.1105]])

Does Tensor A equal Tensor B? (anywhere)


tensor([[False, False, False, False],
        [False, False, False, False],
        [False, False, False, False]])

Wie Sie vielleicht erwartet haben, haben die Tensoren unterschiedliche Werte. Aber was wäre, wenn Sie zwei zufällige Tensoren mit den *gleichen* Werten erstellen wollten? Das heißt, die Tensoren würden immer noch zufällige Werte enthalten, aber sie wären vom gleichen Geschmack.

Hier kommt [`torch.manual_seed(seed)`](https://pytorch.org/docs/stable/generated/torch.manual_seed.html) ins Spiel, wobei `seed` eine ganze Zahl ist (z.B. `42`, aber es kann alles sein), die den Zufallswert bestimmt. Probieren wir es aus, indem wir ein paar weitere *verfälschte* Zufallstensoren erzeugen.



In [ ]:
import torch
import random

# # Set the random seed
RANDOM_SEED = 42  # try changing this to different values and see what happens to the numbers below
torch.manual_seed(seed=RANDOM_SEED)
random_tensor_C = torch.rand(3, 4)

# Have to reset the seed every time a new rand() is called
# Without this, tensor_D would be different to tensor_C
torch.random.manual_seed(
    seed=RANDOM_SEED
)  # try commenting this line out and seeing what happens
random_tensor_D = torch.rand(3, 4)

print(f"Tensor C:\n{random_tensor_C}\n")
print(f"Tensor D:\n{random_tensor_D}\n")
print(f"Does Tensor C equal Tensor D? (anywhere)")
random_tensor_C == random_tensor_D


Tensor C:
tensor([[0.8823, 0.9150, 0.3829, 0.9593],
        [0.3904, 0.6009, 0.2566, 0.7936],
        [0.9408, 0.1332, 0.9346, 0.5936]])

Tensor D:
tensor([[0.8823, 0.9150, 0.3829, 0.9593],
        [0.3904, 0.6009, 0.2566, 0.7936],
        [0.9408, 0.1332, 0.9346, 0.5936]])

Does Tensor C equal Tensor D? (anywhere)


tensor([[True, True, True, True],
        [True, True, True, True],
        [True, True, True, True]])

Sehr schön!

Es sieht so aus, als ob das Setzen des Seeds funktioniert hat.

**Ressourcen:** Was wir gerade behandelt haben, kratzt nur an der Oberfläche der Reproduzierbarkeit in PyTorch. Für mehr, über Reproduzierbarkeit im Allgemeinen und zufällige Seeds, würde ich mir das anschauen:
> * [Die PyTorch-Dokumentation zur Reproduzierbarkeit] (https://pytorch.org/docs/stable/notes/randomness.html) (eine gute Übung wäre es, diese 10 Minuten lang durchzulesen, und selbst wenn Sie es jetzt nicht verstehen, ist es wichtig, es zu wissen).
> * [The Wikipedia random seed page](https://en.wikipedia.org/wiki/Random_seed) (dies gibt einen guten Überblick über Zufallssamen und Pseudozufälligkeit im Allgemeinen).



## Tensoren auf GPUs ausführen (und schnellere Berechnungen durchführen)

Deep-Learning-Algorithmen erfordern eine Menge numerischer Operationen.

Und standardmäßig werden diese Operationen oft auf einer CPU (Computer Processing Unit) durchgeführt.

Es gibt jedoch noch eine andere gängige Hardware, die Grafikprozessoreinheit (GPU), die bei der Ausführung der spezifischen Arten von Operationen, die neuronale Netze benötigen (Matrixmultiplikationen), oft viel schneller ist als CPUs.

Vielleicht verfügt Ihr Computer über eine solche Einheit.

Wenn ja, sollten Sie ihn so oft wie möglich zum Trainieren von neuronalen Netzen verwenden, da sich die Trainingszeit dadurch mit großer Wahrscheinlichkeit drastisch verkürzt.

Es gibt mehrere Möglichkeiten, erstens Zugang zu einem Grafikprozessor zu bekommen und zweitens PyTorch dazu zu bringen, den Grafikprozessor zu nutzen.

**Hinweis:** Wenn ich in diesem Kurs von einem "Grafikprozessor" spreche, beziehe ich mich auf einen [Nvidia-GPU mit CUDA](https://developer.nvidia.com/cuda-gpus), der aktiviert ist (CUDA ist eine Computerplattform und API, die es ermöglicht, Grafikprozessoren für allgemeine Berechnungen und nicht nur für Grafiken zu verwenden), sofern nicht anders angegeben.




### 1. Eine GPU besorgen

Wenn ich GPU sage, wissen Sie vielleicht schon, worum es geht. Aber wenn nicht, gibt es einige Möglichkeiten, um an eine zu kommen.

| **Methode** | **Schwierigkeit der Einrichtung** | **Vorteile** | **Nachteile** | **Wie man einrichtet** |
| ----- | ----- | ----- | ----- | ----- |
| Google Colab | Einfach | Kostenlos zu verwenden, fast keine Einrichtung erforderlich, kann Arbeit mit anderen teilen, so einfach wie ein Link | Speichert Ihre Datenausgaben nicht, begrenzte Rechenleistung, unterliegt Zeitüberschreitungen | [Folgen Sie der Google Colab-Anleitung](https://colab.research.google.com/notebooks/gpu.ipynb) |
| Verwenden Sie Ihr eigenes System | Mittel | Führen Sie alles lokal auf Ihrem eigenen Rechner aus | GPUs sind nicht kostenlos, erfordern Vorabkosten | Befolgen Sie die [PyTorch Installationsrichtlinien](https://pytorch.org/get-started/locally/) |
| Cloud Computing (AWS, GCP, Azure) | Mittelschwer | Geringe Vorabkosten, Zugang zu fast unendlichen Rechenkapazitäten | Kann bei Dauerbetrieb teuer werden, benötigt einige Zeit für die richtige Einrichtung | Befolgen Sie die [PyTorch-Installationsrichtlinien](https://pytorch.org/get-started/cloud-partners/) |

Es gibt noch weitere Optionen für die Nutzung von GPUs, aber die drei oben genannten reichen vorerst aus.

Ich persönlich verwende eine Kombination aus Google Colab und meinem eigenen Computer für kleinere Experimente (und die Erstellung dieses Kurses) und greife auf Cloud-Ressourcen zurück, wenn ich mehr Rechenleistung benötige.

**Ressourcen:** Wenn Sie einen eigenen Grafikprozessor kaufen möchten, aber nicht sicher sind, was Sie kaufen sollen, [Tim Dettmers hat einen hervorragenden Leitfaden](https://timdettmers.com/2020/09/07/which-gpu-for-deep-learning/).

Um zu überprüfen, ob Sie Zugang zu einem Nvidia-Grafikprozessor haben, können Sie `!nvidia-smi` ausführen, wobei das `!` (auch bang genannt) bedeutet: "Führen Sie dies auf der Kommandozeile aus".



In [ ]:
!nvidia-smi


Der Befehl "nvidia-smi" ist entweder falsch geschrieben oder
konnte nicht gefunden werden.


Wenn Sie keinen Nvidia-Grafikprozessor zur Verfügung haben, wird die obige Ausgabe etwa so aussehen:

```
NVIDIA-SMI ist fehlgeschlagen, weil es nicht mit dem NVIDIA-Treiber kommunizieren konnte. Stellen Sie sicher, dass der neueste NVIDIA-Treiber installiert ist und läuft.
```

In diesem Fall gehen Sie wieder nach oben und folgen Sie den Installationsschritten.

Wenn Sie einen Grafikprozessor haben, wird die obige Zeile etwas wie folgt ausgeben:


```
Wed Jan 19 22:09:08 2022       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 495.46       Driver Version: 460.32.03    CUDA Version: 11.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla P100-PCIE...  Off  | 00000000:00:04.0 Off |                    0 |
| N/A   35C    P0    27W / 250W |      0MiB / 16280MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+
```

### 2. PyTorch auf der GPU zum Laufen bringen

Sobald Sie eine GPU zur Verfügung haben, besteht der nächste Schritt darin, PyTorch für die Speicherung von Daten (Tensoren) und die Berechnung von Daten (Durchführung von Operationen auf Tensoren) zu verwenden.

Dazu können Sie das Paket [`torch.cuda`](https://pytorch.org/docs/stable/cuda.html) verwenden.

Anstatt darüber zu reden, sollten wir es lieber ausprobieren.

Sie können testen, ob PyTorch Zugang zu einer GPU hat, indem Sie [`torch.cuda.is_available()`](https://pytorch.org/docs/stable/generated/torch.cuda.is_available.html#torch.cuda.is_available) verwenden.



In [ ]:
# Check for GPU
import torch

torch.cuda.is_available()


False

Wenn das obige Ergebnis `True` ist, kann PyTorch die GPU sehen und benutzen, wenn es `False` ist, kann es die GPU nicht sehen und in diesem Fall müssen Sie die Installationsschritte erneut durchführen.

Nehmen wir nun an, Sie möchten Ihren Code so einrichten, dass er auf der CPU *oder* der GPU läuft, wenn diese verfügbar ist.

Auf diese Weise funktioniert Ihr Code, wenn Sie oder jemand anderes ihn ausführen möchte, unabhängig von dem verwendeten Computergerät.

Erstellen wir eine Variable `device`, um zu speichern, welche Art von Gerät verfügbar ist.


In [ ]:
# Set device type
device = "cuda" if torch.cuda.is_available() else "cpu"
device


'cpu'

Wenn die obige Ausgabe "cuda" bedeutet, dass wir unseren gesamten PyTorch-Code so einstellen können, dass er das verfügbare CUDA-Gerät (eine GPU) verwendet, und wenn sie "cpu" ausgibt, bleibt unser PyTorch-Code bei der CPU.

**Hinweis:** In PyTorch ist es die beste Praxis, [**Geräteunabhängigen Code**] (https://pytorch.org/docs/master/notes/cuda.html#device-agnostic-code) zu schreiben. Das bedeutet Code, der auf der CPU (immer verfügbar) oder der GPU (falls verfügbar) läuft.

Wenn Sie schneller rechnen wollen, können Sie eine GPU verwenden, aber wenn Sie *viel* schneller rechnen wollen, können Sie mehrere GPUs verwenden.

Sie können die Anzahl der GPUs, auf die PyTorch Zugriff hat, mit [`torch.cuda.device_count()`](https://pytorch.org/docs/stable/generated/torch.cuda.device_count.html#torch.cuda.device_count) zählen.


In [ ]:
# Count number of devices
torch.cuda.device_count()


0

Die Kenntnis der Anzahl der GPUs, auf die PyTorch Zugriff hat, ist hilfreich, wenn Sie einen bestimmten Prozess auf einer GPU und einen anderen Prozess auf einer anderen GPU laufen lassen wollen (PyTorch hat auch Funktionen, mit denen Sie einen Prozess auf *allen* GPUs laufen lassen können).

### 2.1 PyTorch zum Laufen bringen auf Apple Silicon

Um PyTorch auf Apples M1/M2/M3-GPUs laufen zu lassen, können Sie das Modul [`torch.backends.mps`](https://pytorch.org/docs/stable/notes/mps.html) verwenden.
- Stellen Sie sicher, dass die Versionen von macOS und Pytorch aktualisiert sind.
- Sie können testen, ob PyTorch Zugriff auf eine GPU hat, indem Sie `torch.backends.mps.is_available()` verwenden.


In [ ]:
# Check for Apple Silicon GPU
import torch

torch.backends.mps.is_available()  # Note this will print false if you're not running on a Mac


False

In [ ]:
# Set device type
device = "mps" if torch.backends.mps.is_available() else "cpu"
device


'cpu'

Wie zuvor bedeutet die obige Ausgabe "mps", dass wir unseren gesamten PyTorch-Code so einstellen können, dass er die verfügbare Apple Silicon GPU verwendet.

In [ ]:
if torch.cuda.is_available():
    device = "cuda"  # Use NVIDIA GPU (if available)
elif torch.backends.mps.is_available():
    device = "mps"  # Use Apple Silicon GPU (if available)
else:
    device = "cpu"  # Default to CPU if no GPU is available


### 3. Tensoren (und Modelle) auf der GPU platzieren

Sie können Tensoren (und Modelle, das werden wir später sehen) auf ein bestimmtes Gerät legen, indem Sie [`to(device)`](https://pytorch.org/docs/stable/generated/torch.Tensor.to.html) aufrufen. Wobei `Gerät` das Zielgerät ist, auf das Sie den Tensor (oder das Modell) legen möchten.

Warum dies tun?

GPUs bieten viel schnellere numerische Berechnungen als CPUs, und wenn keine GPU verfügbar ist, wird der Code aufgrund unseres **geräteunabhängigen Codes** (siehe oben) auf der CPU ausgeführt.

**Hinweis:** Wenn Sie einen Tensor auf die GPU bringen, indem Sie `to(device)` verwenden (z.B. `some_tensor.to(device)`), wird eine Kopie dieses Tensors zurückgegeben, d.h. derselbe Tensor wird auf CPU und GPU sein. Um Tensoren zu überschreiben, ordnen Sie sie neu zu:
>
> "irgendein_tensor = irgendein_tensor.to(device)`

Versuchen wir, einen Tensor zu erstellen und ihn auf der GPU zu platzieren (wenn sie verfügbar ist).



In [ ]:
# Create tensor (default on CPU)
tensor = torch.tensor([1, 2, 3])

# Tensor not on GPU
print(tensor, tensor.device)

# Move tensor to GPU (if available)
tensor_on_gpu = tensor.to(device)
tensor_on_gpu


tensor([1, 2, 3]) cpu


tensor([1, 2, 3])

Wenn Sie einen Grafikprozessor zur Verfügung haben, wird der obige Code etwas Ähnliches ausgeben:

```
tensor([1, 2, 3]) cpu
tensor([1, 2, 3], device='cuda:0')
```

Beachten Sie, dass der zweite Tensor `device='cuda:0'` hat, was bedeutet, dass er auf der 0. verfügbaren GPU gespeichert ist (GPUs sind mit 0 indiziert, wenn zwei GPUs verfügbar wären, wären sie `'cuda:0'` bzw. `'cuda:1'`, bis zu `'cuda:n'`).



### 4. Verschieben von Tensoren zurück in die CPU

Was wäre, wenn wir den Tensor zurück zur CPU verschieben wollten? Dies ist zum Beispiel sinnvoll, wenn Sie mit NumPy mit Ihren Tensoren interagieren wollen (NumPy nutzt die GPU nicht). Versuchen wir, die Methode [`torch.Tensor.numpy()`](https://pytorch.org/docs/stable/generated/torch.Tensor.numpy.html) für unseren "Tensor_on_gpu" zu verwenden.



In [ ]:
# If tensor is on GPU, can't transform it to NumPy (this will error)
tensor_on_gpu.numpy()


array([1, 2, 3], dtype=int64)

Stattdessen können wir [`Tensor.cpu()`](https://pytorch.org/docs/stable/generated/torch.Tensor.cpu.html) verwenden, um einen Tensor zurück auf die CPU zu bringen und mit NumPy zu verwenden.

Dies kopiert den Tensor in den CPU-Speicher, so dass er mit CPUs verwendet werden kann.

In [ ]:
# Instead, copy the tensor back to cpu
tensor_back_on_cpu = tensor_on_gpu.cpu().numpy()
tensor_back_on_cpu


array([1, 2, 3], dtype=int64)

Die obige Funktion gibt eine Kopie des GPU-Tensors im CPU-Speicher zurück, so dass sich der ursprüngliche Tensor immer noch auf der GPU befindet.

In [ ]:
tensor_on_gpu


tensor([1, 2, 3])

## Übungen

Alle Übungen konzentrieren sich auf das Einüben des obigen Codes.

Sie sollten in der Lage sein, sie zu lösen, indem Sie sich auf die einzelnen Abschnitte beziehen oder indem Sie die verlinkten Ressourcen befolgen.

**Ressourcen:**

* [Übungsvorlagen-Notizbuch für 00](https://github.com/mrdbourke/pytorch-deep-learning/blob/main/extras/exercises/00_pytorch_fundamentals_exercises.ipynb).
* [Beispiellösungsheft für 00](https://github.com/mrdbourke/pytorch-deep-learning/blob/main/extras/solutions/00_pytorch_fundamentals_exercise_solutions.ipynb) (versuchen Sie die Übungen *bevor* Sie sich dies ansehen).

1. Lesen der Dokumentation - Ein großer Teil des Deep Learning (und des Programmierens im Allgemeinen) besteht darin, sich mit der Dokumentation eines bestimmten Frameworks vertraut zu machen, das Sie verwenden. Wir werden die PyTorch-Dokumentation im weiteren Verlauf dieses Kurses häufig verwenden. Ich würde also empfehlen, sich 10 Minuten Zeit zu nehmen, um das Folgende zu lesen (es ist in Ordnung, wenn Sie einige Dinge vorerst nicht verstehen, der Schwerpunkt liegt noch nicht auf dem vollständigen Verständnis, sondern auf dem Bewusstsein). Siehe die Dokumentation zu [`torch.Tensor`](https://pytorch.org/docs/stable/tensors.html#torch-tensor) und zu [`torch.cuda`](https://pytorch.org/docs/master/notes/cuda.html#cuda-semantics).
2. Erzeugen Sie einen zufälligen Tensor mit der Form `(7, 7)`.
3. Führen Sie eine Matrixmultiplikation auf dem Tensor von 2 mit einem anderen zufälligen Tensor mit der Form `(1, 7)` durch (Hinweis: Sie müssen den zweiten Tensor eventuell transponieren).
4. Setzen Sie den Zufallskeim auf `0` und wiederholen Sie die Übungen 2 & 3.
5. Apropos Zufallssaat, wir haben gesehen, wie man sie mit `torch.manual_seed()` setzt, aber gibt es ein GPU-Äquivalent? (Hinweis: Sie müssen dazu in die Dokumentation von `torch.cuda` schauen). Wenn ja, setzen Sie den GPU-Zufallswert auf "1234".
6. Erzeugen Sie zwei zufällige Tensoren der Form `(2, 3)` und senden Sie beide an die GPU (Sie brauchen dafür Zugang zu einer GPU). Setzen Sie `torch.manual_seed(1234)` beim Erstellen der Tensoren (dies muss nicht der GPU-Zufallswert sein).
7. Führen Sie eine Matrixmultiplikation mit den Tensoren durch, die Sie in 6 erstellt haben (auch hier müssen Sie möglicherweise die Form eines der Tensoren anpassen).
8. Finde den maximalen und minimalen Wert der Ausgabe von 7.
9. Ermitteln Sie den maximalen und minimalen Indexwert der Ausgabe von 7.
10. Bilden Sie einen zufälligen Tensor mit der Form `(1, 1, 1, 10)` und erstellen Sie dann einen neuen Tensor, bei dem alle `1` Dimensionen entfernt wurden, um einen Tensor mit der Form `(10)` zu erhalten. Setze den Seed auf `7` wenn du ihn erstellst und drucke den ersten Tensor und seine Form sowie den zweiten Tensor und seine Form aus.



## Extra-Lehrplan

* Verbringen Sie 1 Stunde damit, das [PyTorch Grundlagen-Tutorial] (https://pytorch.org/tutorials/beginner/basics/intro.html) durchzuarbeiten (ich empfehle die Abschnitte [Quickstart] (https://pytorch.org/tutorials/beginner/basics/quickstart_tutorial.html) und [Tensoren] (https://pytorch.org/tutorials/beginner/basics/tensorqs_tutorial.html)).
* Um mehr darüber zu erfahren, wie ein Tensor Daten darstellen kann, sehen Sie sich dieses Video an: [Was ist ein Tensor?](https://youtu.be/f5liqUk0ZTw)

